Add purchase flow for WooCommerce.com-hosted plugins

yscik opened the pull request:
Part of #3139

### Changes proposed in this Pull Request

* When selecting any paid extension, open a new tab for WooCommerce.com checkout with the cart with the paid extensions added
* Meanwhile continue installing the rest of the extensions from Wordpress.org
* Remove the feature flag `setup_wizard_all_extensions` that hides these extensions by default.

### Testing instructions

* An internet-accessible test site is required
* Install Sensei and access the Setup Wizard features step `/wp-admin/admin.php?page=sensei_setup_wizard&step=features`
* Select a paid plugin, click `Continue` and `Install now`
* A new tab should open with a WooCommerce.com checkout screen for the selected plugin.
* Complete the purchase.
* Click `Add to Site` button on the order summary screen.
* Select the site to install the plugin to, or add a new site and go trough the process, connecting and authorizing the site to WooCommerce.com
* The plugin should be installed and activated on the site.
* Back on the the setup wizard tab, it should show the plugin as installed. 
--
* For easier testing of checkout, add this to `includes/admin/class-sensei-setup-wizard.php#L582` to make `Sensei LMS Certificates` purchased and installed via the WC.com flow without any need for payment shenanigans.
  ```
	if ( 'sensei-certificates' === $extension->product_slug ) {
					$extension->price = '$0.00';
					$extension->wccom_product_id = 247548;
				}
  ```

yscik commented:
Some notes @donnapep 

* Looks like a prompt to connect the site to WC.com is offered as part of the installation flow, so it should work out with a newly installed WooCommerce plugin.
* I think there is a high chance that with this flow the user will not get back to the setup wizard - only on the happy path of purchasing the plugins and successfully finding and going through `Add to site` will this happen. 
* On that note, and because it's also simpler technically, we could just start installing the WP.org plugins immediately, not waiting for the WC.com flow with them. Those installations can go on in the background. 
* We could also open the WC.com checkout in a new tab
* The explanation that WooCommerce is needed, and mainly the part explaining they will be redirected to purchase those extensions could be a bit more prominent - I'll put up a screenshot with a notice box to try.

jom commented on includes/admin/class-sensei-setup-wizard.php:
```diff
@@ -417,17 +417,51 @@ public function get_mailing_list_form_data() {
+			'admin.php'
+		);
+
+		return [
```
Do you think it'd be worth doing a method check and pulling these directly from `\WC_Admin_Addons::get_in_app_purchase_url_params`? I guess it is a gamble of what is most likely to break/change: the method name (hopefully deprecated) or their implementation for what they pass to WCCOM.

yscik replied:
Nice find, added a branch for this in 7430ab1. The previous code is still there as a fallback so we can prepare the parameters in case WooCommerce is not yet installed.

jom commented on includes/admin/class-sensei-setup-wizard.php:
```diff
@@ -417,17 +417,51 @@ public function get_mailing_list_form_data() {
+
+		$is_woocommerce_installed = Sensei_Utils::is_plugin_present_and_activated( 'Woocommerce', 'woocommerce/woocommerce.php' );
+
+		if ( ! $is_woocommerce_installed ) {
```
We might also want to do a version check. Some quick research, I think this functionality was added in WC 3.7.0, but not 100% sure.

yscik replied:
Added this too above, just for the the `WC_Admin_Addons` usage.
Looks like the auto-install helper itself was integrated around 3.1, not sure what minimum is needed for current WooCommerce.com compatibilty. But the checkout will work without any of that.

yscik commented:
In case the user has WooCommerce version `< 3.7.0` installed, we can redirect them to the WC.com checkout but they will have to download and install the plugin zips manually after purchase.

donnapep commented:
> On that note, and because it's also simpler technically, we could just start installing the WP.org plugins immediately, not waiting for the WC.com flow with them. Those installations can go on in the background.

What I was hoping to avoid was the user seeing in-progress spinners for the free extensions, before they were interrupted and taken to WC.com. However, if we open WC.com in a new tab, that becomes a non-issue, so I think this would be fine.

> I'll put up a screenshot with a notice box to try.

I'd rather not show that in a notice. That may look a bit scary for beginners and could cause them to change their mind about installing, which wouldn't be good. Besides that, if we're now opening WC.com in a new tab, it seems as though it would be less of an issue since they won't be suddenly redirected away.

To summarize:
- Free extensions can be installed at the same time as WC.
- After WC is successfully installed, WC.com opens in a new tab. This will also enable them to get back to the setup wizard in case of an install error with a paid extension.

donnapep commented:
Thinking more about the "Once WooCommerce is installed..." disclaimer, I think we could even remove it from the feedback screen and just leave it on the modal. On the modal, we could move it down and add it as a separate paragraph below the "You won't have access to this functionality..." and maybe bold it or something?

donnapep commented:
Further to this idea of opening in a new tab, is it going to make it more difficult to detect when the extension was successfully installed or encountered an error? That will be needed for showing the retry links and enabling the _Continue_ button.

More importantly though, we'll still want to log events like `sensei_setup_wizard_features_install_success` and `sensei_setup_wizard_features_install_error`. Otherwise, we'll only be able to capture their intent to install a paid plugin, but won't be able to measure if they were successful at it.

yscik commented:
We can keep polling for status changes on the features screen, it should be up-to date with the installation progress regardless of it being done on a separate tab/background processes.

yscik commented:
Some issues and changes that came up:
* Opening the WooCommerce.com checkout as soon as the user clicks `Install now`.
  New tabs can only be opened in browsers without popup blocking via user-initated events (like a click), so we cannot wait for Woocommerce to be installed. The good news is we don't have to, as it's not required to initiate the checkout. By the time the user finishes the purchase, it should be installed and auto-install can work.
* After a successful purchase and install, it's not possible to make WooCommerce.com redirect to the Setup Wizard directly. It only uses the site path from the `wccom-back` parameter, and redirects to the `plugins.php` page. Since we now open the checkout in a new tab the Setup Wizard is not lost so this should be fine.
  * A possible workaround used in Woo-admin is checking for a WooCommerce.com checkout referer header and redirecting if a Sensei plugin was installed. 
* When purchasing multiple extensions, after installing the first via`Add to Site`, the user is redirected to our site, without a way back to auto-install the other extension. 
* We should consider adding a `Skip` to ready step option at the `Installing` screen, as this now waits and blocks the Setup wizard if the user does not complete the purchase.

renatho commented on assets/setup-wizard/features/index.js:
```diff
@@ -106,19 +116,14 @@ const Features = () => {
 			setSelectedSlugs( ( prev ) => [ ...prev, wcSlug ] );
 		}
-	}, [ getSelectedFeatures, features ] );
+	}, [ getSelectedFeatures, features, isWooCommerceInstalled ] );
```
I think now we could remove the `features` from here, right?

yscik replied:
Yup removed in d61d5a3

renatho commented on assets/setup-wizard/features/index.js:
```diff
@@ -106,19 +116,14 @@ const Features = () => {
 			return;
 		}
 
-		const wooCommerceFeature = features.find( ( f ) => wcSlug === f.slug );
```
Good idea separating it! =)

yscik commented:
>  if there is an error while installing the WooCommerce

As an alternative, we can add a `Purchase` button to open the checkout that is shown in the feedback screen only after WooCommerce is successfully installed, and keep the current flow only when WooCommerce was already installed

> retry the installation in case of errors 

If the `Add to Site` auto-install fails WooCommerce.com will indicate it with a try again option and some links to support. But it all stays on the WooCommerce.com site. 

I'm not sure we can reliably detect various errors in this without extensive workarounds and hooks into the whole WC.com connect process.

yscik commented on assets/setup-wizard/features/features-selection.js:
```diff
@@ -63,7 +63,7 @@ const FeaturesSelection = ( {
 							onChange={ toggleItem( slug ) }
 							checked={ selectedSlugs.includes( slug ) }
-							disabled={ !! status }
+							disabled={ 'installed' === status }
```
Sorry about that, this was just leftover from when I tried out a new status. Removed in 56be8fd

yscik commented on assets/setup-wizard/helpers/woocommerce-com.js:
```diff
@@ -0,0 +1,28 @@
+ *
+ * @return {string} The product ID.
+ */
+export const getWcProductId = ( feature ) => feature.wccom_product_id;
```
👍 e092175

renatho commented:
> As an alternative, we can add a Purchase button to open the checkout that is shown in the feedback screen only after WooCommerce is successfully installed, and keep the current flow only when WooCommerce was already installed.
> 
> I think the user can still go through with the purchase in this case, and worst case they will install the plugins manually.

I remember some time @donnapep talking that she wanted to prevent additional user actions. Maybe we should only consider as fallback the plugin manually installation. Is it intuitive when the WC plugin installation fails? Or weird things can happen?

> If the Add to Site auto-install fails WooCommerce.com will indicate it with a try again option and some links to support. But it all stays on the WooCommerce.com site.

Oh, good! I didn't know that! IMHO, if the WC.com already handles that, we don't need to worry about it. Maybe, just make sure that nothing weird happens when the user comes back from WC.com in all the possible situations.

renatho commented:
I tested now trying to install the plugin in a Jurassic Ninja site. And I think it didn't recognize my site. Is it because it's running in a new tab? I think it should automatically identify the site when adding the parameters `wccom-site` and `wccom-back`.

jom commented:
I was just about to post the same thing. I tried this on Atomic Ephemeral and it doesn't recognize my site when I go through the checkout flow. I manually pasted in the URL to add the site and it connected fine. I think went back and clicked Install. WC.com never recognized it was installed (just kept spinning) but the plugins were installed (I had to do this individually from the order confirmation screen for both plugins). When I went to do it for WCPC, it installed (again, WC.com kept spinning), but Sensei never recognized WCPC being installed (althought in Plugins it appeared installed/activated). It did recognize Content Drip as activated in the Setup Wizard feature progress.

yscik commented:
* Fixed site connect flow
* Added an external status to wccom plugins, instead of the installing spinner. Installation does not wait for these so the user can continue to the ready step, regardless of the WC.com purchase outcome:

<img width="538" alt="image" src="https://user-images.githubusercontent.com/176949/85453869-9ce4fe00-b59c-11ea-9a63-4f67f0a4f3d2.png">

jom commented:
Code looks good and works well. I'll let @renatho give the final thumbs-up.

Just a note: Products appear when they can't be installed by the installed version of WooCommerce, but I don't think that is a huge problem. It looks like ~28% of WC sites are on a version less than 3.7. I think WooCommerce.com already hides the feature to install on a site based on the version passed, but they can still buy and install manualy. I think especially with the ability to move on in the setup wizard without the WCCOM plugins installed we should be fine.
